### ISSUE TITLE
Wireguard module issue / upgrade 18.09 -> 19.03

### ISSUE THREAD

**Issue by eonpatapon** [NixOS/nixpkgs#56040]
## Issue description

Upgraded my system to 19.03. Wireguard service fails with:

```
Feb 19 10:30:30 fion systemd[1]: Starting WireGuard Tunnel - wg0...
Feb 19 10:30:32 fion vpsgyw8mg48k1fa9r8f4ym5031mkr683-unit-script-wireguard-wg0-start[3620]: modprobe: FATAL: Module wireguard not found in directory /run/booted-system/kernel-modules/lib/modules/4.14.88
Feb 19 10:30:32 fion systemd[1]: wireguard-wg0.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 10:30:32 fion 9agw5bcpff1nr2amdwhvnwpd2sid64jv-unit-script-wireguard-wg0-post-stop[3680]: Cannot find device "wg0"
Feb 19 10:30:32 fion systemd[1]: wireguard-wg0.service: Control process exited, code=exited status=1
Feb 19 10:30:32 fion systemd[1]: wireguard-wg0.service: Failed with result 'exit-code'.
Feb 19 10:30:32 fion systemd[1]: Failed to start WireGuard Tunnel - wg0.

```
This is strange since wireguard module is correctly loaded with my current kernel:

```
[root@fion:~]# uname -a
Linux fion 4.14.88 #1-NixOS SMP Thu Dec 13 08:18:54 UTC 2018 x86_64 GNU/Linux

[root@fion:~]# lsmod | grep wireguard
wireguard             225280  0
ip6_udp_tunnel         16384  1 wireguard
udp_tunnel             16384  1 wireguard
ipv6                  483328  70 nf_conntrack_ipv6,ip6t_rpfilter,nf_defrag_ipv6,wireguard
```

However the module is missing somehow:

```
[root@fion:~]# find -L /run/booted-system/kernel-modules/lib/modules/4.14.88/ -name '*wireguard*'
```

But it's available with the new kernel from 19.03:
```
[root@fion:~]# find -L /run/current-system/kernel-modules/lib/modules/4.14.99/ -name '*wireguard*'
/run/current-system/kernel-modules/lib/modules/4.14.99/extra/wireguard.ko.xz
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.88, NixOS, 19.03.git.020e024 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by dasJ** [comment]
Do you need to reboot to apply a kernel update? Compare the kernel versions:
```
file -L /run/current-system/kernel
```
and
```
uname -r
```

**Comment by eonpatapon** [comment]
Didn't reboot yet but I guess it will work.

Still it's strange that the module is removed from the current kernel installation

**Comment by arianvp** [comment]
Is this related to https://github.com/NixOS/nixpkgs/issues/51773 ?

**Comment by kirelagin** [comment]
I am seeing the same thing. The module is neither in `/run/booted-system/kernel-modules/`, nor in `/run/current-system/kernel-modules/`, but `lsmod` shows that it is loaded 😕.

**Comment by kirelagin** [comment]
Ah, correction: it _is_ in `/run/current-system`, I just failed to use `find` with `-L`. But not in `/run/booted-system`.

**Comment by kirelagin** [comment]
Ok, this is starting to look like it is not related to the NixOS upgrade: our `booted-system` genuinely did not have wireguard enabled. That’s confusing, and I don’t understand how it worked before, but now I believe, all we really need is a reboot.

**Comment by kirelagin** [comment]
Ok, I figured it. So, it is not related specifically to the 18.09 -> 19.03 upgrade, it can happen with any upgrade of NixOS that upgrades the kernel.

`modprobe` is looking for kernel modules in two places:

1. `/run/current-system/kernel-modules/lib/modules/$(uname -r)`
2. `/run/booted-system/kernel-modules/lib/modules/$(uname -r)`

If your `booted-system` did not have wireguard (or any other kernel module actually) enabled, you changed the configuration _and_ the kernel version remained the same, `modprobe` will find the new module in location 1 and load it. But if the kernel gets upgraded later, location 1 will no longer work due to the kernel version mismatch, and location 2 will not work either, so you have to reboot.

**Comment by asymmetric** [comment]
> location 1 will no longer work due to the kernel version mismatch

What do you mean "no longer work"? Woudn't location 1 point to the new kernel, which supposedly has the module?

**Comment by kirelagin** [comment]
Yeah, sorry, I think my explanation was a bit cumbersome (I imagine at that time I have spent all my mental power on figuring out the cause 😁).

So, this involves three steps:

1. Boot into a system that has no wireguard.
2. Enable wireguard in your config without upgrading NixOS (alternatively: on a newer version of NixOS but one that comes with the same kernel version).
    Now you have no module in `.../booted-system/.../$(uname -r)/` but the module gets loaded from `.../current-system/.../$(uname -r)/`.
3. Upgrade NixOS to a version that ships with a newer kernel.
    You still have no module in `.../booted-system/.../$(uname -r)/` (it was never there in the first place) _and_ now there is no module in `.../current-system/.../$(uname -r)/` (since `uname -r` is now different from the version of the kernel in `current-system`).

@asymmetric what you are saying is also true, in your case steps 2 and 3 just essentially get squashed into one. What makes this particular issue so subtle is that in the 3-step process between steps 2 and 3 you actually have working wireguard and the module is actually loaded, even though it no longer exists at the paths that `modeprobe` tries.

**Comment by asymmetric** [comment]
Right. I worked around my particular situation by pinning the kernel, enablling wireguard, then unpinning and proceeding like normal.

**Comment by PaulGrandperrin** [comment]
Hi, I ran into this issue, but without a kernel upgrade, I just tried to install `sysdig` using `programs.sysdig.enable`.
Once installed, the module doesn't load:

```
$ modprobe sysdig
modprobe: FATAL: Module sysdig not found in directory /run/booted-system/kernel-modules/lib/modules/5.15.19
```

However, the module is available at `/run/current-system/kernel-modules/lib/modules/$(uname -r)`:

```
$ file /run/current-system/kernel-modules/lib/modules/(uname -r)/extra/sysdig-probe.ko.xz
/run/current-system/kernel-modules/lib/modules/5.15.19/extra/sysdig-probe.ko.xz: symbolic link to /nix/store/61zfsgrhkrad9fhyk3hmw5np561ghyww-sysdig-0.27.1/lib/modules/5.15.19/extra/sysdig-probe.ko.xz

$ file /run/booted-system/kernel-modules/lib/modules/(uname -r)/extra/sysdig-probe.ko.xz
/run/booted-system/kernel-modules/lib/modules/5.15.19/extra/sysdig-probe.ko.xz: cannot open `/run/booted-system/kernel-modules/lib/modules/5.15.19/extra/sysdig-probe.ko.xz' (No such file or directory)
```

It seems that `modprobe` is not searching at `/run/current-system/kernel-modules/lib/modules/$(uname -r)` despite https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/os-specific/linux/kmod/default.nix#L7

`strace` seems to confirm that:

```
$ strace -f -e trace=file modprobe sysdig 2>&1|rg '(current|booted)-system'
execve("/run/current-system/sw/bin/modprobe", ["modprobe", "sysdig"], 0x7fffef6ae6e0 /* 43 vars */) = 0
newfstatat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19", {st_mode=S_IFDIR|0555, st_size=17, ...}, 0) = 0
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.softdep", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.dep.bin", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.alias.bin", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.symbols.bin", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.builtin.alias.bin", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/run/booted-system/kernel-modules/lib/modules/5.15.19/modules.builtin.bin", O_RDONLY|O_CLOEXEC) = 3
modprobe: FATAL: Module sysdig not found in directory /run/booted-system/kernel-modules/lib/modules/5.15.19

```

Feels like `modprobe` is not honoring its `--with-modulesdirs` build flag.

Any ideas about why?

**Comment by WolfangAukang** [comment]
Same is happening with a PR merging 'vmware-host' module. It is using [kmod's modprobe](https://github.com/NixOS/nixpkgs/blob/764618d46cf8f82cc01fc759b5bab931626f5bb2/nixos/modules/virtualisation/vmware-host.nix#L153) for a certain service, and it is complaining about not finding the modules, as shown [here](https://github.com/NixOS/nixpkgs/pull/169966#issuecomment-1112598319).

It is looking at `/run/booted-system' instead of `/run/current-system`, where the modules are located.

**Comment by WolfangAukang** [comment]
Pinging kmod's maintainer @Artturin (I don't see any mentions on this issue)

**Comment by Artturin** [comment]
> It is looking at `/run/booted-system` instead of `/run/current-system`, where the modules are located.

i made it look in `/run/booted-system`first in https://github.com/NixOS/nixpkgs/pull/149244 (end of 2021)


if something is not found in `/run/booted-system` then it should look in `/run/current-system` so maybe thats not working

**Comment by WolfangAukang** [comment]
> if something is not found in /run/booted-system then it should look in /run/current-system so maybe thats not working

Indeed, my wording on the previous comment is incorrect. Although modprobe should be looking at `/run/booted-system` and `/run/current-system`, my guess (I don't know how kmod works) is that it "crashes" when not finding the modules on the first path.

**Comment by Artturin** [comment]
yeah, maybe the patch stopped working properly at some point as the latest update to it was in 2016
